File: 백준/Silver/1463. 1로 만들기/1로 만들기.py
def solve(N):
    i = 2
    while i <= N:
        candidates = []
        if i % 3 == 0:
            candidates.append(memo[i//3])
        if i % 2 == 0:
            candidates.append(memo[i//2])
        candidates.append(memo[i-1])
        memo[i] = 1+min(candidates)
        i += 1
    return memo[N]

N = int(input())

memo = [0] * (N+1)
print(solve(N))

File: 백준/Silver/1463. 1로 만들기/1로 만들기.java
import java.util.Arrays;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        var scanner = new Scanner(System.in);
        var n = Integer.parseInt(scanner.nextLine());

        System.out.println(solve(n));
    }
    
    static int solve(int x) {
        int[] memo = new int[x + 1];
        Arrays.fill(memo, Integer.MAX_VALUE);
        memo[0] = memo[1] = 0;
        
        for (int i = 2; i < x + 1; i++) {
            if (i % 3 == 0) {
                memo[i] = Math.min(memo[i], memo[i / 3] + 1);
            }
            if (i % 2 == 0) {
                memo[i] = Math.min(memo[i], memo[i / 2] + 1);
            }
            memo[i] = Math.min(memo[i], memo[i - 1] + 1);
        }
        return memo[x];
    }
}
